sqlglot/optimizer/simplify.py

```diff
@@ -547,7 +547,29 @@ def simplify_literals(expression, root=True):
 NULL_OK = (exp.NullSafeEQ, exp.NullSafeNEQ, exp.PropertyEQ)
 
 
+def _simplify_integer_cast(expr: exp.Expression) -> exp.Expression:
+    if isinstance(expr, exp.Cast) and expr.this.is_int:
+        num = int(expr.this.name)
+        # Remove the (up)cast from small (byte-sized) integers in predicates which is side-effect free. Downcasts on any
+        # integer type might cause overflow, thus the cast cannot be eliminated and the behavior is
+        # engine-dependent
+        if (
+            exp.DataType.TINYINT_MIN <= num <= exp.DataType.TINYINT_MAX
```

Comment:
How hard would it be to support all upcasts in one go?

Reply:
For context, we discussed doing this in one go, but it may be a bit risky because the type widths aren't necessarily global for integer types. Handling a single byte was a conservative choice that seemed the most safe. The alternative would be to ensure that the widths are the same or control them per dialect so that the optimizer won't silently "fix" broken queries that would overflow if the data were corrupted.

Reply:
Makes sense